File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/tools/ToolDisplay.java
package wraith.fabricaeexnihilo.compatibility.rei.tools;

import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.display.Display;
import me.shedaniel.rei.api.common.entry.EntryIngredient;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.compatibility.rei.ReiIngredientUtil;
import wraith.fabricaeexnihilo.recipe.ToolRecipe;
import wraith.fabricaeexnihilo.recipe.util.BlockIngredient;

import java.util.List;
import java.util.Optional;

//TODO: Make this like sieve, big box of outputs
public class ToolDisplay implements Display {
    public final CategoryIdentifier<ToolDisplay> category;
    public final EntryIngredient block;
    public final EntryIngredient result;
    private final Identifier id;

    public ToolDisplay(ToolRecipe recipe, CategoryIdentifier<ToolDisplay> category) {
        this.category = category;
        BlockIngredient blockIngredient = recipe.getBlock();
        this.block = ReiIngredientUtil.of(blockIngredient);
        this.result = EntryIngredients.of(recipe.getResult().stack());
        this.id = recipe.getId();
    }

    @Override
    public CategoryIdentifier<ToolDisplay> getCategoryIdentifier() {
        return category;
    }

    @Override
    public List<EntryIngredient> getInputEntries() {
        return List.of(block);
    }

    @Override
    public List<EntryIngredient> getOutputEntries() {
        return List.of(result);
    }

    @Override
    public Optional<Identifier> getDisplayLocation() {
        return Optional.of(id);
    }
}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/tools/ToolDisplay.kt
package exnihilofabrico.compatibility.rei.tools

import exnihilofabrico.api.recipes.ToolRecipe
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.modules.ModTools
import exnihilofabrico.util.asREIEntry
import me.shedaniel.rei.api.EntryStack
import me.shedaniel.rei.api.RecipeDisplay
import net.minecraft.util.Identifier

class ToolDisplay(val recipe: ToolRecipe, val category: Identifier): RecipeDisplay {
    override fun getRecipeCategory() = category

    override fun getOutputEntries() = recipe.lootables.map { it.stack.asREIEntry() }
    override fun getInputEntries(): MutableList<MutableList<EntryStack>>? {
        val ingredients = recipe.ingredient.asREIEntries()
        val tools =
            (if(recipeCategory == PluginEntry.HAMMER) ModTools.HAMMERS else ModTools.CROOKS)
                .values.map{ it.asREIEntry() }.toMutableList()

        return mutableListOf(ingredients, tools)
    }

}